services/oauth/v2/utils.go

```diff
@@ -0,0 +1,157 @@
+		"errorMessage":    s.errorMessage,
+		"isCallToCpApi":   strconv.FormatBool(s.isCallToCpApi),
+		"authErrCategory": s.authErrCategory,
+		"destType":        s.destDefName,
+		"isTokenFetch":    strconv.FormatBool(s.isTokenFetch),
+		"flowType":        string(s.flowType),
+		"action":          s.action,
+	}
+	stats.Default.NewTaggedStat(s.statName, stats.CountType, statsTags).Increment()
+}
+
+// {input: [{}]}
+// {input: [{}, {}, {}, {}]}
+// {input: [{}, {}, {}, {}]} -> {output: [{200}, {200}, {401,authErr}, {401,authErr}]}
+func GetAuthErrorCategoryFromTransformResponse(respData []byte) (string, error) {
+	var transformedJobs []TransformerResponse
+	err := jsonfast.Unmarshal([]byte(gjson.GetBytes(respData, "output").Raw), &transformedJobs)
+	if err != nil {
+		return "", err
+	}
+	tfJob, found := lo.Find(transformedJobs, func(item TransformerResponse) bool {
+		return IsValidAuthErrorCategory(item.AuthErrorCategory)
+	})
+	if !found {
+		// can be a valid scenario
+		return "", nil
+	}
+	return tfJob.AuthErrorCategory, nil
+}
+
+func GetAuthErrorCategoryFromTransformProxyResponse(respData []byte) (string, error) {
+	transformedJobs := TransformerResponse{}
+	err := jsonfast.Unmarshal([]byte(gjson.GetBytes(respData, "output").Raw), &transformedJobs)
+	if err != nil {
+		return "", err
+	}
+	return transformedJobs.AuthErrorCategory, nil
+}
+
+func checkIfTokenExpired(secret AccountSecret, oldSecret json.RawMessage, expiryTimeDiff time.Duration, stats *OAuthStats) bool {
+	expirationDate := expirationDate{}
+	if err := json.Unmarshal(secret.Secret, &expirationDate); err != nil {
+		stats.errorMessage = "unmarshal failed"
+		stats.statName = GetOAuthActionStatName("proActive_token_refresh")
+		stats.SendCountStat()
+		return false
+	}
+	if expirationDate.ExpirationDate != "" && isTokenExpired(expirationDate.ExpirationDate, expiryTimeDiff, stats) {
+		return true
+	}
+	if router_utils.IsNotEmptyString(string(oldSecret)) {
+		if bytes.Equal(secret.Secret, oldSecret) {
+			return true
+		}
+	}
+	return false
+}
+
+func isTokenExpired(expirationDate string, expirationTimeDiff time.Duration, stats *OAuthStats) bool {
+	date, err := time.Parse(misc.RFC3339Milli, expirationDate)
+	if err != nil {
+		stats.errorMessage = "parsing failed"
+		stats.statName = GetOAuthActionStatName("proActive_token_refresh")
+		stats.SendCountStat()
+		return false
+	}
+	if date.Before(time.Now().Add(expirationTimeDiff)) {
+		return true
+	}
+	return false
+}
+
+func GetErrorType(err error) string {
+	errTypMap := map[syscall.Errno]string{
+		syscall.ECONNRESET:   "econnreset",
+		syscall.ECONNREFUSED: "econnrefused",
+		syscall.ECONNABORTED: "econnaborted",
+		syscall.ECANCELED:    "ecanceled",
+	}
+	if os.IsTimeout(err) {
+		return "timeout"
+	}
+	for errno, errTyp := range errTypMap {
+		if ok := errors.Is(err, errno); ok {
+			return errTyp
+		}
+	}
+	var e net.Error
+	if errors.As(err, &e) {
+		return "network_error"
+	}
+	return "none"
+}
```

Comment:
```suggestion
func GetErrorType(err error) string {
	if os.IsTimeout(err) {
		return "timeout"
	}
	if lo.Contains(lo.Keys(ErrTypMap), err) {
		return ErrTypMap[err]
	}
	var e net.Error
	if errors.As(err, &e) {
		return "network_error"
	}
	return "none"
}
```

Reply:
We can't do this. Getting this error `type error of err does not match inferred type syscall.Errno for T`.

Reply:
Can we resolve this @cisse21 ?